src/mesh/p4est_mesh.jl

```diff
@@ -117,16 +117,103 @@ function P4estMesh(trees_per_dimension; polydeg,
+      boundary_names[1, tree] = :x_neg
+
+      tree = linear_indices[end, cell_y]
+      boundary_names[2, tree] = :x_pos
+    end
+  end
+
+  # Boundaries in y-direction
+  if !periodicity[2]
+    for cell_x in 1:trees_per_dimension[1]
+      tree = linear_indices[cell_x, 1]
+      boundary_names[3, tree] = :y_neg
+
+      tree = linear_indices[cell_x, end]
+      boundary_names[4, tree] = :y_pos
+    end
+  end
+
+  return P4estMesh{NDIMS, RealT, NDIMS+2}(p4est, p4est_mesh, tree_node_coordinates, nodes,
+                                          boundary_names, "", unsaved_changes)
 end
 
 
-# Create a new p4est_connectivity that represents a structured rectangle with periodic boundaries.
-# Similar to p4est_connectivity_new_brick, but doesn't use Morton ordering.
-# This ordering makes `calc_node_coordinates!` below easier but is irrelevant otherwise.
-# TODO P4EST non-periodic
-function connectivity_structured_periodic(cells_x, cells_y)
+"""
+    P4estMesh(meshfile::String;
+              mapping=nothing, polydeg=1, RealT=Float64,
+              initial_refinement_level=0, unsaved_changes=true)
+
+Import an uncurved, unstructured, conforming mesh from an Abaqus mesh file (`.inp`),
+map the mesh with the specified mapping, and create a `P4estMesh` from the curved mesh.
+
+Cells in the mesh file will be imported as trees in the `P4estMesh`.
+
+# Arguments
+- `meshfile::String`: an uncurved Abaqus mesh file that can be imported by p4est.
+- `mapping`: a function of `NDIMS` variables to describe the mapping that transforms
+             the imported mesh to the physical domain. Use `nothing` for the identity map.
+- `polydeg::Integer`: polynomial degree used to store the geometry of the mesh.
+                      The mapping will be approximated by an interpolation polynomial
+                      of the specified degree for each tree.
+                      The default of `1` creates an uncurved geometry. Use a higher value if the mapping
+                      will curve the imported uncurved mesh.
+- `RealT::Type`: the type that should be used for coordinates.
+- `initial_refinement_level::Integer`: refine the mesh uniformly to this level before the simulation starts.
+- `unsaved_changes::Bool`: if set to `true`, the mesh will be saved to a mesh file.
+"""
+function P4estMesh(meshfile::String;
+                   mapping=nothing, polydeg=1, RealT=Float64,
+                   initial_refinement_level=0, unsaved_changes=true)
+  # TODO P4EST
+  NDIMS = 2
+
+  # Prevent p4est from crashing Julia if the file doesn't exist
+  @assert isfile(meshfile)
+
+  conn = p4est_connectivity_read_inp(meshfile)
+
+  # These need to be of the type Int for unsafe_wrap below to work
+  n_trees::Int = conn.num_trees
+  n_vertices::Int = conn.num_vertices
+
+  vertices        = unsafe_wrap(Array, conn.vertices, (3, n_vertices))
+  tree_to_vertex  = unsafe_wrap(Array, conn.tree_to_vertex, (4, n_trees))
+
+  basis = LobattoLegendreBasis(RealT, polydeg)
```

Comment:
Maybe add a TODO comment (an issue?) that mentions that this hardcodes the use of the `LobattoLegendreBasis` for this mesh type and thus also enforces the use of the same basis for the solver, doesn't it? Or are these two unrelated?

Reply:
These two are unrelated. The geometry of the trees is saved at `nodes` and will be interpolated to whatever the solver is using. The solver can even have a higher polydeg (but not a lower one because this will cause non-conforming meshes to not be "watertight" anymore).

Reply:
Got it. Thanks!